presto-main/src/main/java/com/facebook/presto/sql/planner/CanonicalPlanGenerator.java

```diff
@@ -54,23 +60,71 @@
 {
     private final PlanNodeIdAllocator planNodeidAllocator = new PlanNodeIdAllocator();
     private final PlanVariableAllocator variableAllocator = new PlanVariableAllocator();
+    private final PlanCanonicalizationStrategy strategy;
+
+    public CanonicalPlanGenerator(PlanCanonicalizationStrategy strategy)
+    {
+        this.strategy = strategy;
+    }
 
     public static Optional<CanonicalPlanFragment> generateCanonicalPlan(PlanNode root, PartitioningScheme partitioningScheme)
     {
         Map<VariableReferenceExpression, VariableReferenceExpression> originalToNewVariableNames = new HashMap<>();
-        Optional<PlanNode> canonicalPlan = root.accept(new CanonicalPlanGenerator(), originalToNewVariableNames);
+        Optional<PlanNode> canonicalPlan = root.accept(new CanonicalPlanGenerator(PlanCanonicalizationStrategy.DEFAULT), originalToNewVariableNames);
         if (!originalToNewVariableNames.keySet().containsAll(partitioningScheme.getOutputLayout())) {
             return Optional.empty();
         }
         return canonicalPlan.map(planNode -> new CanonicalPlanFragment(planNode, getCanonicalPartitioningScheme(partitioningScheme, originalToNewVariableNames)));
     }
 
+    public static Optional<CanonicalPlan> generateCanonicalPlan(PlanNode root, PlanCanonicalizationStrategy strategy)
+    {
+        Optional<PlanNode> canonicalPlanNode = root.accept(new CanonicalPlanGenerator(strategy), new HashMap<>());
+        if (!canonicalPlanNode.isPresent()) {
+            return Optional.empty();
+        }
+        return Optional.of(new CanonicalPlan(canonicalPlanNode.get(), strategy));
+    }
+
     @Override
     public Optional<PlanNode> visitPlan(PlanNode node, Map<VariableReferenceExpression, VariableReferenceExpression> context)
     {
         return Optional.empty();
     }
 
+    @Override
+    public Optional<PlanNode> visitOutput(OutputNode node, Map<VariableReferenceExpression, VariableReferenceExpression> context)
```

Comment:
Is this needed for connector strategy as well?

Reply:
This is needed for all strategies except DEFAULT which is used for FRC.